### ISSUE TITLE
[docs] Search finds duplicates

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#4954]
Searching for `torch.nn.functional.conv2d` returns results with 4 identical links to http://pytorch.org/docs/master/nn.html?highlight=torch%20nn%20functional%20conv2d#torch.nn.functional.conv2d

Link with results is: http://pytorch.org/docs/master/search.html?q=torch.nn.functional.conv2d&check_keywords=yes&area=default

This is true for master and 0.3

**Comment by vadimkantorov** [comment]
@zou3519

**Comment by vadimkantorov** [comment]
@vishwakftw Could you please label this as `documentation`? Thanks!

**Comment by vadimkantorov** [comment]
Also https://pytorch.org/docs/stable/search.html?q=nn.gru&check_keywords=yes&area=default duplicates torch.nn.GRU

**Comment by vadimkantorov** [comment]
Another instance with bincount:
![image](https://user-images.githubusercontent.com/1041752/101241062-4b1bd480-36f3-11eb-9a09-bbb049ab6e46.png)

I would say some postproc of search results should de-duplicate function pages. Multiple occurences of search term there don't deserve to be separate search hits

**Comment by vadimkantorov** [comment]
Same for `clip_grad_value_`:
![image](https://user-images.githubusercontent.com/1041752/107324203-f0627880-6aa7-11eb-8641-7c9e0745fb50.png)

**Comment by vadimkantorov** [comment]
Search also has quite bad snippets. Parts of code examples are most often useless as snippets
![image](https://user-images.githubusercontent.com/1041752/190831005-a09d892d-a5e2-4d08-a9fa-1830dcefd16b.png)

Same applied to removing code examples from docs pages available for indexing by Google

**Comment by vadimkantorov** [comment]
If the query is matches exactly or very near symbol names, search should return excellent snippets (preferably even function signatures directly)

**Comment by vadimkantorov** [comment]
@malfet @ezyang The docs website search relevance seems broken:

Here is what `nn.linear` search returns:

<img width="2894" height="1638" alt="Image" src="https://github.com/user-attachments/assets/3d6649cc-beba-458b-8796-2c67f7f58106" />

Plenty of duplicates, and a lot of various stub quantized variants displayed before `torch.nn.Linear`

**Comment by malfet** [comment]
Lots of things have changed in the last 9 years... for example there is no more docs team. But we have agents that should figure out a few solutions for Python docs search.

I'm going to try a few options here and see what gives the best relevance for symbol-name queries: Pagefind, Lunr.js and FlexSearch.

For context on the current state: the docs site actually ships two search paths. The default is Google Programmable Search, with Google's suggest endpoint powering the typeahead; behind a toggle there's still stock Sphinx client-side search. Neither dedupes the ~26 `Linear` object entries in the index, which is why the quantized stubs crowd out `torch.nn.Linear`.

**Comment by vadimkantorov** [comment]
> ~26 `Linear` object entries in the index

Also, in search results are duplicates. Could be fixed with some postprocessing of symbol-like results

**Comment by vadimkantorov** [comment]
And if relevance is hard to tweak, maybe somehow boost up results of torch.nn, torch.nn.functional and hard-code other such important namespaces, and put down all other more esoteric ones

**Comment by albanD** [comment]
I would also highly recommend enabling the little ticker in the search bar. I uses google's system for indexing our website and searching. Which makes it a lot better.

**Comment by vadimkantorov** [comment]
> Another instance with `torch.bincount`:

Regarding this, I wonder if some "canonicalization" can happen as well, e.g. when there are module method and instance method, merge them into one snippet (as probably dozens or hundred of functions exist in both version and clutter search results), e.g. dedup can group / take into account canonical symbol name (if it's present)

Like so, maybe even some sort of grouping for dozens of variants of Linear can happen (but this can be more controversial)

**Comment by vadimkantorov** [comment]
Seems still not merged, for 2.14 docs

**Comment by vadimkantorov** [comment]
Symbol duplicates still remain (`torch.add`), despite theme commit message mentioning that it eliminated duplicates. Snippets are still quite trashy as well :(

If it's using pagefind, pagefind allows adding HTML attributes to mark the part of HTML which needs to be indexed, so that nav is not indexed

<img width="1742" height="1390" alt="Image" src="https://github.com/user-attachments/assets/fcc7658a-5155-424b-b90a-094351d0cc56" />
